### DISCUSSION TITLE
Automate backups & exports hourly

### DISCUSSION THREAD

**Discussion by Mag1cByt3s** [outline/outline#6553] [Self Hosting]
I just had an issue with the database after upgrading.

I had no success in fixing the errors.

Thankfully i still had a backup but that was a few hours old already.

Since my documents are very mission-critical i would like to automatically backup everything every hour.

So far I only know the export function which I can trigger manually.

I would like to automate that and save it to my storage server. I could easily mount a remote nfs path to save the backups but i have no idea how i can automatically request and save the export.

Any help with that would be appreciated.

Thanks.

**Comment by tommoor** [comment]
The best way to backup on such a regular basis would be to take a snapshot of the server disk, this is a process that would be done completely outside of the Outline software. Depending on your host it may have such capability built into the server management.

**Reply by Mag1cByt3s** [reply]
I am currently taking full snapshots of the RAID array every 24 hours.

Also i am taking backups of the outltine folder including database every hour via an rsync cronjob.

Still the database was corrupted even after replacing the backup it could only be restored with the full backup i had on TrueNAS. But that was 24 hours old.

**Reply by tommoor** [reply]
It would be interesting to get more details about the database corruption. That sounds like a pretty rare scenario..

**Reply by Mag1cByt3s** [reply]
I don't have logs about these errors anymore.

But it complained about columns in the database being already existend as it tried to run the migration script on startup after the update.

I tried to edit the db manually by deleting these columns but that just caused more issues.

**Comment by Mag1cByt3s** [comment]
So small update to this: My database got just nuked again. I just opened outline and my data was completely gone.
I have no clue what happened i only saw authorization errors in the logs.
After copying over a postgres backup my documents are back but I assume that some data is lost.
The problem is that i have so many notes that i am not sure what is lost.

I need a reliable way to backup everything.

But at least backing up the postgres folder seems to be enough if the db version did not change.

**Reply by tommoor** [reply]
> My database got just nuked again

It sounds like your postgres container storage is not setup correctly, you can get behavior similar to this if using Docker without mapping the data directory to a volume.

**Reply by Mag1cByt3s** [reply]
Here is my current docker-compose stack: https://privatebin.net/?e3e5fb0b2b3836f4#BXN89x76FsTxF8c1qUoG6crLHeyZXpp8i2sU6AEEceJD

maybe someone can have a look. 

The mount point /mnt/nas/outline is mounted via NFS from TrueNAS.

**Comment by Mag1cByt3s** [comment]
I am using TrueNAS to store the DB on a RAID1 system. I wonder if i can snapshot the postgres folder every 30 minutes or so.

I am also now considering taking off-site backups.

Otherwise i might have to look into using the API to export my collections: https://www.getoutline.com/developers#tag/Collections/paths/~1collections.export/post

It would be interesting to see how other people handle the backup strategy here.

My current backup strategy are just a few cronjobs:

```
# create an hourly backup of outline
0 * * * * /usr/bin/rsync -a /mnt/pve/truenas/outline/ /mnt/pve/truenas/Backups/outline.hourly >/dev/null 2>&1
0 * * * * /usr/bin/rsync -a /mnt/pve/truenas/outline/ /papa/storage/Backups/Pascal/outline.hourly >/dev/null 2>&1

# create a half-hourly backup of outline
*/30 * * * * /usr/bin/rsync -a /mnt/pve/truenas/outline/ /mnt/pve/truenas/Backups/outline.halfhourly >/dev/null 2>&1
```

TrueNAS is snapshotting and backuping the whole RAID every 24 hours to a backup drive.

**Comment by Mag1cByt3s** [comment]
While looking through the issues here on Github i stumbled upon this: https://github.com/captn3m0/outliner
That seems to be exactly what i need.

**Reply by tommoor** [reply]
I would just warn that it looks like that library hasn't been updated in 4 years, the API is _mostly_ compatible over that time but there have also been many changes as you can imagine.

**Reply by Mag1cByt3s** [reply]
True. I also noticed that. I was hoping it would still be compatible. I may have to build something on my own that talks to the API and downloads an export.

**Comment by ReessKennedy** [comment]
+1 - Beyond the scenario for redundancy, I use Obsidian for markdown notes and if Outline were configure to run hourly backups of markdown files to GoogleDrive I could sync these with my personal Obsidian vault and then I could know that when I search my laptop locally or search inside my Obsidian vault, I'd have access to all my knowledge - both my personal, kept in plain markdown files in Obsidian, and my team's stuff hosted remotely and managed by Outline.

**Reply by erikboberg** [reply]
I agree with this! Would fit my use case perfectly.

**Reply by tommoor** [reply]
Guys this is what the API is for - automate away!

**Comment by lrstanley** [comment]
In case anyone else is curious, I had the same need, exporting outline docs, both as a general backup, but also to replicate to github so I can still read them if my entire Kubernetes clusters where I'm hosting it is offline.

I've made **outline-export** for this purpose: https://github.com/lrstanley/outline-export -- it can both automate the zip export from the UI in whatever format you'd like, but it can also extract, auto-sanitize, files/folders, if you wanted to do something like export to markdown and push to a git repo.

Just needs the URL to your Outline instance, and an appropriately scoped API key. Binaries and docker images available, if you want to run as a cron in a containerized environment (like I will be, using a `CronJob` in Kubernetes).
